File: libc/kheap.c
#include <kheap.h>

static uint32_t _kmalloc (uint32_t size, uint32_t align); 
static void fill (uint32_t end, uint32_t size, uint8_t value);

uint8_t *bitset_base = (uint8_t*) BITSET_BASE;
uint8_t *bitset_limit = (uint8_t*) BITSET_LIMIT;
uint8_t *heap_base = (uint8_t*) HEAP_BASE;
uint32_t kheap_enable = 0;

uint32_t kmalloc_u (uint32_t sz)
{
  return _kmalloc (sz, 0);
}

uint32_t kmalloc (uint32_t sz)
{
  return _kmalloc (sz, 1);
}

void kheap_init ()
{
  kheap_enable = 1;
  memset ((uint8_t*)bitset_base, 0 ,bitset_limit - bitset_base);
}
 
static uint32_t _kmalloc (uint32_t size, uint32_t align)
{
  uint8_t *i;
  uint32_t count = 0;
  uint32_t end = 0;
  for (i = bitset_base; i < bitset_limit ; i++)
  {
    uint32_t num_bits_from_base = (((uint32_t)i) - BITSET_BASE) * 8;
    if (align && !count && (num_bits_from_base % 0x1000) && num_bits_from_base){
      end += 8;
      continue;
    }
    uint32_t j;
    for (j = 0; j < 8; j++)
    {
      end++;
      if ((*i >> j) & 1)
        count = 0;
      else
        count++;
      if (count == size)
      {
        fill (end - size, size, 1);
        return (uint32_t)((((((uint32_t)i) - BITSET_BASE) * 8) + (j + 1)) + HEAP_BASE) - size;
      }
    }
  }
  return 0;
}

static void fill (uint32_t begin, uint32_t size, uint8_t value)
{
  uint32_t i;
  for (i = begin; i < begin + size; i++) 
  {
    uint32_t byte = i / 8;
    uint32_t offset = i % 8;

    if (value)
      *(bitset_base + byte) |= (1 << (offset)); 
    else
      *(bitset_base + byte) &= ~(1 << (offset)); 
  }

}

void kfree (uint32_t addr, uint32_t size)
{
  uint32_t byte = (uint32_t)addr;
  uint32_t tmp = (byte - HEAP_BASE);
  fill (tmp, size, 0);
}

void print_bit_map (uint32_t size)
{
  size = (size / 8) + (size % 8 != 0);
  uint8_t *i = bitset_limit;
  for (i = bitset_base ; i < (bitset_base + size); i++)
  {
    int8_t j;
    for (j = 0; j < 8; j++)
    {
      if ((*i >> j) & 1)
        kprint_debug ("1", CYAN);
      else
        kprint_debug ("0", CYAN);
    }
  }
}

File: src/tests/libc/kheap.py
import gdb
import os
import sys
sys.path.append(os.path.dirname('..'))
import james_utils

utils = james_utils.JamesUtils()
utils.GdbInit()

## TEST KHEAP.h

### KHEAP_INIT
gdb.execute('call kheap_init()')
kheap_enable = gdb.parse_and_eval('kheap_enable')
assert(kheap_enable == 0x1)

gdb_bitset_base = gdb.parse_and_eval('bitset_base')
bitset_base = int.from_bytes(gdb_bitset_base.bytes, byteorder = 'little', signed = False)
gdb_bitset_limit = gdb.parse_and_eval('bitset_limit')
bitset_limit = int.from_bytes(gdb_bitset_limit.bytes, byteorder = 'little', signed = False)
expected = b'\x00' * (bitset_limit - bitset_base)
assert(utils.TestMemoryRegion(expected, bitset_base))
print('\033[92m[+] KHEAP iniciado corretamente.\033[0m')

### FILL
gdb.execute('call fill(0, 12, 1)')
expected = b'\xff\x0f'
gdb_bitset_base = gdb.parse_and_eval('bitset_base')
bitset_base = int.from_bytes(gdb_bitset_base.bytes, byteorder = 'little', signed = False)
assert(utils.TestMemoryRegion(expected, bitset_base))
print('\033[92m[+] KHEAP fill(0, 12,1) executado corretamente.\033[0m')

gdb.execute('call fill(4, 4, 0)')
expected = b'\x0f\x0f'
gdb_bitset_base = gdb.parse_and_eval('bitset_base')
bitset_base = int.from_bytes(gdb_bitset_base.bytes, byteorder = 'little', signed = False)
assert(utils.TestMemoryRegion(expected, bitset_base))
print('\033[92m[+] KHEAP fill(4, 4, 0) executado corretamente.\033[0m')

gdb.execute('call fill(0, 12, 0)')
expected = b'\x00\x00'
gdb_bitset_base = gdb.parse_and_eval('bitset_base')
bitset_base = int.from_bytes(gdb_bitset_base.bytes, byteorder = 'little', signed = False)
assert(utils.TestMemoryRegion(expected, bitset_base))
print('\033[92m[+] KHEAP fill(0, 0, 0) executado corretamente.\033[0m')

### KMALLOC
kmalloc = gdb.parse_and_eval('_kmalloc(0, 0)')
assert(kmalloc == 0x0)
print('\033[92m[+] KHEAP kmalloc(0, 0) executado corretamente.\033[0m')

kmalloc = gdb.parse_and_eval('_kmalloc(1, 0)')
gdb_heap_base = gdb.parse_and_eval('heap_base')
heap_base = int.from_bytes(gdb_heap_base.bytes, byteorder = 'little', signed = False)
assert(kmalloc == heap_base)
print('\033[92m[+] KHEAP kmalloc(1, 0) executado corretamente.\033[0m')

kmalloc = gdb.parse_and_eval('_kmalloc(1, 1)')
gdb_heap_base = gdb.parse_and_eval('heap_base')
heap_base = int.from_bytes(gdb_heap_base.bytes, byteorder = 'little', signed = False)
try:
    assert(kmalloc == heap_base + 0x1000)
    print('\033[92m[+] KHEAP kmalloc(1, 1) executado corretamente.\033[0m')
except:
    print('\033[31m[-] KHEAP kmalloc(1, 1) executado com erro.\033[0m')

gdb_heap_base = gdb.parse_and_eval('heap_base')
heap_base = int.from_bytes(gdb_heap_base.bytes, byteorder = 'little', signed = False)
gdb.execute(f'call kfree({heap_base}, 1)')
try:
    expected = b'\x02' # 00000010
    gdb_bitset_base = gdb.parse_and_eval('bitset_base')
    bitset_base = int.from_bytes(gdb_bitset_base.bytes, byteorder = 'little', signed = False)
    assert(utils.TestMemoryRegion(expected, bitset_base))
    print(f'\033[92m[+] KHEAP kfree({heap_base}, 1) executado corretamente.\033[0m')
except:
    print(f'\033[31m[-] KHEAP kfree({heap_base}, 1) executado com erro.\033[0m')

utils.GdbStop()
